### PR TITLE
feat: add a crate for encoders and decoders to eventually be used as part of v2 reader/writer

### PR DESCRIPTION
**Pull request by westonpace**
This PR is quite large and still quite incomplete.

Notable TODOs:

 * Lots more testing.  Need to test all data types, test nulls, multiple pages, staggered pages, empty writes, etc.
 * Add support for simple struct encoding / decoding.  By "simple struct" I mean "can never be null".  In this case we only need to write a header column with a single page and no buffers.  The header column could technically be omitted but, in the further future, I expect there to be multiple ways to encode structs, and the header column will be needed to identify which mechanism is being used.  Structs with nulls are a more complex feature altogether.
 * Add null support to lists. The .proto file actually describes the approach I would like to take (u64 offsets with bitpack-friendly sentinels).  However, this isn't implemented yet.
 * Add support / tests for fixed size list (this should be a pretty straightforward "value encoder" but the field mapping might not be working correct yet)

However, I believe I now have the overall shape of things correct.  I would like to both checkpoint my work and solicit feedback.  Plus, if this gets any bigger it will be even more impossible to review.

### CODE REVIEW

File: rust/lance-encoding/src/encoder.rs [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+    pub is_data: bool,
+    /// Buffers that make up the encoded buffer
+    ///
+    /// All of these buffers should be written to the file as one contiguous buffer
+    ///
+    /// This is a Vec to allow for zero-copy
+    ///
+    /// For example, if we are asked to write 3 primitive arrays of 1000 rows and we can write them all
+    /// as one page then this will be the value buffers from the 3 primitive arrays
+    pub parts: Vec<Buffer>,
```
Comment (wjones127) [inline rust/lance-encoding/src/encoder.rs:35]: IMO "chunks" might be a better name than "parts" here.

**Review by wjones127 — COMMENTED** [review]
Some initial comments. This design seems quite good.

File: protos/encodings.proto [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    AllNull all_nulls = 3;
+  }
+}
+
+// An array encoding for variable-length list fields
+message List {
+    // An array containing the offsets into the items array.
+    //
+    // This array will have (num_rows + 1), will have a data
+    // type of uint64, and will never have nulls.
```
[the remark this answers is not in the document]
Comment (westonpace) [reply protos/encodings.proto:126]: So a few things.  First,  I haven't implemented what I described here.  So I need to get back to that.

> Never have nulls? null behavior is described right below though, right?

The list array has nulls.  The offsets array will never have nulls (nulls in the list are "past the max" offsets)

> Also, why uint64 over signed int64? If we do signed then at least in one case we are more aligned with Arrow. Also easier on Java later on if that ever becomes a factor.

I was worried that signed integers wouldn't bitpack well.  However, there are strategies, like [zigzag](https://docs.rs/residua-zigzag/latest/zigzag/) to re-encode signed integers in a scheme that isn't 2's complement, which would probably work well if we still want to use signed integers.  So I guess this isn't too much of a concern.

Signed integers would also make the explanation easier (if an item is null then encode -1 * the previous offset)

File: protos/encodings.proto [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    //
+    // offsets[0] will always be 0.
+    //
+    // If the incoming list at index i is not null then offsets[i+1]
+    // will contain offsets[i] + len(list)
+    //
+    // If the incoming list at index i is null then offsets[i+1] will
+    // contain offsets[i] + num_items
```
[the remark this answers is not in the document]
Comment (westonpace) [reply protos/encodings.proto:134]: No, I was assuming this field would be restricted to 63 bits of actual values.

File: rust/lance-encoding/src/decoder.rs [hunk 4]
```diff
@@ -0,0 +1,733 @@
...
+            .into_iter()
+            .map(|col_tasks| {
+                let arrays = col_tasks
+                    .into_iter()
+                    .map(|col_task| col_task.decode())
+                    .collect::<Result<Vec<_>>>()?;
+                let array_refs = arrays.iter().map(|arr| arr.as_ref()).collect::<Vec<_>>();
+                // TODO: If this is a primtiive column we should be able to avoid this
+                // allocation + copy with "page bridging" which could save us a few CPU
+                // cycles.
```
[the remark this answers is not in the document]
Comment (westonpace) [reply rust/lance-encoding/src/decoder.rs:508]: End-to-end zero copy would only be available in the case where we can't use any kind of encoding at all.  However, that may be the case for our FSL vectors since embeddings are pretty compressed already.

Otherwise, some kind of buffer copy is going to be inevitable because the shorter encoded representation is going to have to be decoded into a longer buffer.  Trying to do that in place is probably more work than its worth.

Either way, we can measure and see.  This is something we can iterate on if decoding starts becoming a CPU bottleneck.
